### DISCUSSION TITLE
the need for sink

### DISCUSSION THREAD

**Discussion by wtholliday** [hylo-lang/hylo#775] [Q&A] [answered]
Just a quick question to check my understanding. A sink parameter is pass-by-move (ownership transfer). It seems to me this could be inferred (either statically or dynamically) for a let parameter, but that would result in potential implicit copies. Val (like Rust) wants to make every copy explicit. Is all that correct?

**Comment by kyouko-taiga** [comment, marked as answer]
Indeed, all copies must be explicit in Val, including for "simple" data types like `Int`.

You can't infer transfer of ownership statically without an interpocedural analysis and doing it dynamically would incur prohibitive runtime costs.

**Reply by wtholliday** [reply]
Regarding dynamic inference of ownership: it seems to me that objects fall roughly into two categories. POD-ish things that are cheap to copy and you can easily pass them around by value (3d vectors, etc... Copy types in Rust), or heavier things where having a do-i-own-this boolean passed along wouldn't be costly compared to actually copying the object (Rust's Vec, Rc, etc). OTOH, I suppose if you are calling a function in a hot loop and passing some array (something costly to copy), you don't want that function you're calling to have to check if it owns the array passed in and thus needs to free it. Does that sound right? Just trying to understand where the prohibitively expensive part comes from.

**Reply by dabrahams** [reply]
Yes that's right.  Just reading a thing should not have any abstraction penalty.  Even checking for a unique reference count requires an atomic operation.

**Comment by dabrahams** [comment]
I should clarify though that while a sink parameter is indeed pass-by-ownership transfer, there is no move operation involved.
